File: PadraoGof/Criacional/FactoryMethod/Factory_method01.js
/* 

--------------------------------------------------------------------------------------------------
Fatec Diadema - Luigi Papaiz 
Curso : Desenvolvimento de Software Multiplataforma - DSM 
Disciplina: Técnica de Programação II
Nome: Kaiki Kenji       
Data: 23/02/2024
Descrição: Aplicação do padrão GoF - Criacional - FactoryMethod desenvolvido como exemplo na aula.
--------------------------------------------------------------------------------------------------
*/

/* Classe Base de Veiculos */

class Veiculo {
    constructor(modelo) {
        this.modelo = modelo;
    }

    mostrarDetalhes() {
        console.log(`Modelo: ${this.modelo}`);
    }
}

/* Subclasses de Veiculos */

class Carro extends Veiculo {
    constructor(modelo) {
        super(modelo);
    }
}

class Moto extends Veiculo {
    constructor(modelo) {
        super(modelo);
    }
}

/* Fabrica Abstrata de Veiculos */

class FabricaDeVeiculos {
    criarVeiculo(modelo) {
        throw new Error("O Método criarVeiculo deve ser implementado pela SubClasse");
    }
}

/* Fabrica Concetra de Carros */

class FabricaDeCarros extends FabricaDeVeiculos {
    criarVeiculo(modelo) {
        return new Carro(modelo);
    }
}

/* Fabrica Concetra de Motos */

class FabricaDeMotos extends FabricaDeVeiculos {
    criarVeiculo(modelo) {
        return new Moto(modelo);
    }
}

/* Uso da Fabrica - Interface */

const fabricaDeCarros = new FabricaDeCarros();
const carro = fabricaDeCarros.criarVeiculo('Sedan');
carro.mostrarDetalhes(); // Saida: Modelo: Sedan

const fabricaDeMotos = new FabricaDeMotos();
const moto = fabricaDeMotos.criarVeiculo('Esportiva');
moto.mostrarDetalhes(); // Saida: Modelo: Esportiva

File: PadraoGof/Criacional/FactoryMethod/Factory_method01.py
#--------------------------------------------------------------------------------------------------
#Fatec Diadema - Luigi Papaiz 
#Curso : Desenvolvimento de Software Multiplataforma - DSM 
#Disciplina: Técnica de Programação II
#Nome: Kaiki Kenji       
#Data: 23/02/2024
#Descrição: Aplicação do padrão GoF - Criacional - FactoryMethod desenvolvido como exemplo na aula.
#--------------------------------------------------------------------------------------------------

# Classe Base de Veiculos
class Veiculo:

    #__init__ = construtor
    #self = this
    def __init__(self,modelo):
        self.modelo = modelo

    def mostrarDetalhes(self):
        print(f"modelo: {self.modelo}")

# Subclasses de Veiculos 

class Carro(Veiculo):
    def __init__(self,modelo):
        super().__init__(modelo)

class Moto(Veiculo):
    def __init__(self,modelo):
        super().__init__(modelo)

# Fabrica Abstrata de Veiculos 

class FabricaDeVeiculos:
    def criarVeiculo(self,modelo):
        raise NotImplementedError("O Método criarVeiculo deve ser implementado pela SubClasse")

# Fabrica Concetra de Carros 

class FabricaDeCarro(FabricaDeVeiculos):
    def criarVeiculo(self,modelo):
        return Carro(modelo)
        
# Fabrica Concetra de Motos 

class FabricaDeMoto(FabricaDeVeiculos):
    def criarVeiculo(self,modelo):
        return Moto(modelo)
        
# Uso da Fabrica - Interface 

fabricaDeCarros = FabricaDeCarro()
fabricaDeMotos = FabricaDeMoto()

carro = fabricaDeCarros.criarVeiculo("Sedan")
carro.mostrarDetalhes()

moto =fabricaDeMotos.criarVeiculo("Esportiva")
moto.mostrarDetalhes()